### PR TITLE
Don't throw on unknown JSON input fields

### LINKED ISSUE #1647: deserializeJson throws on unknown input fields
currently (0.7.30) deserializeJson cannot handle input data that it does not expect:

```
struct Test{
  string foo;
}
auto obj = deserializeJson!Test(`{"foo": "42", "bar":""}`);
assert(obj.foo == "42");
```

this throws:
```
std.json.JSONException@../../../../../../.dub/packages/vibe-d-0.7.30/vibe-d/source/vibe/data/json.d(1773): Expecting ',' or '}', not '"'.
----------------
4   alexa-openwebif                     0x0000000102c7dd5b pure @safe bool std.exception.enforceEx!(std.json.JSONException).enforceEx!(bool).enforceEx(bool, lazy immutable(char)[], immutable(char)[], ulong) + 111
5   alexa-openwebif                     0x0000000102c7de10 pure @safe void vibe.data.json.enforceJson!("../../../../../../.dub/packages/vibe-d-0.7.30/vibe-d/source/vibe/data/json.d", 1773uL).enforceJson(bool, lazy immutable(char)[]) + 60
6   alexa-openwebif                     0x0000000102c7d96b void vibe.data.json.JsonStringSerializer!(immutable(char)[], false).JsonStringSerializer.readDictionary!(vibe.data.serialization.__T20deserializeValueImplTS4vibe4data4json34__T20JsonStringSerializerTAyaVbi0Z20JsonStringSerializerS404vibe4data13serialization13DefaultPolicyZ.deserializeValue!(main.main(immutable(char)[][]).Test).deserializeValue(ref vibe.data.json.JsonStringSerializer!(immutable(char)[], false).JsonStringSerializer).Traits).readDictionary(scope void delegate(immutable(char)[])) + 351
7   alexa-openwebif                     0x0000000102c7d6f1 main.main(immutable(char)[][]).Test vibe.data.serialization.__T20deserializeValueImplTS4vibe4data4json34__T20JsonStringSerializerTAyaVbi0Z20JsonStringSerializerS404vibe4data13serialization13DefaultPolicyZ.deserializeValue!(main.main(immutable(char)[][]).Test).deserializeValue(ref vibe.data.json.JsonStringSerializer!(immutable(char)[], false).JsonStringSerializer) + 77
8   alexa-openwebif                     0x0000000102c7d6a0 main.main(immutable(char)[][]).Test vibe.data.serialization.__T21deserializeWithPolicyTS4vibe4data4json34__T20JsonStringSerializerTAyaVbi0Z20JsonStringSerializerS404vibe4data13serialization13DefaultPolicyTS4main4mainFAAyaZ4TestTAyaZ.deserializeWithPolicy(immutable(char)[]) + 60
9   alexa-openwebif                     0x0000000102c7d65b main.main(immutable(char)[][]).Test vibe.data.serialization.deserialize!(vibe.data.json.JsonStringSerializer!(immutable(char)[], false).JsonStringSerializer, main.main(immutable(char)[][]).Test, immutable(char)[]).deserialize(immutable(char)[]) + 31
10  alexa-openwebif                     0x0000000102c7d47b main.main(immutable(char)[][]).Test vibe.data.json.deserializeJson!(main.main(immutable(char)[][]).Test, immutable(char)[]).deserializeJson(immutable(char)[]) + 31
11  alexa-openwebif                     0x0000000102c6aa18 _Dmain + 56
12  alexa-openwebif                     0x0000000102e3f64f D2rt6dmain211_d_run_mainUiPPaPUAAaZiZ6runAllMFZ9__lambda1MFZv + 39
13  alexa-openwebif                     0x0000000102e3f57b void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).tryExec(scope void delegate()) + 35
14  alexa-openwebif                     0x0000000102e3f5f4 void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).runAll() + 44
15  alexa-openwebif                     0x0000000102e3f57b void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).tryExec(scope void delegate()) + 35
16  alexa-openwebif                     0x0000000102e3f4e1 _d_run_main + 497
17  alexa-openwebif                     0x0000000102c6ab39 main + 33
18  libdyld.dylib                       0x00007fff8d1c55ac start + 0
19  ???                                 0x0000000000000000 0x0 + 0
```

### PR DESCRIPTION
**Pull request by HenkKalkwater**
An attempt to fix #1647

~~Still WIP, because of:~~
* [x] ~~All unit test compile and work again (right now, dmd generates cryptic template errors)~~

### CODE REVIEW

**Comment by HenkKalkwater** [comment]
I'm not sure what part the CI build failure caused on `DC=ldc, DVersion=1.16.0, arch=x86`. Is this the fault of my code or just bad luck?

**Comment by s-ludwig** [comment]
> I'm not sure what part the CI build failure caused on DC=ldc, DVersion=1.16.0, arch=x86. Is this the fault of my code or just bad luck?

I haven't seen the error message, but I guess it's a flaky test, there doesn't seem to be anything that looks like it could fail on 32-bit.

**Comment by s-ludwig** [comment]
The change looks good so far, except that it makes a breaking change to the serialization API (there are some serializer implementations outside of vibe.d). I see three possible ways around this: 

1. Let the serialization module select the return value that is accepted by the serializer
2. Introduce `ignoreValue`, or maybe `skipValue` as an optional primitive that is called by the `serialization` code instead of returning `false`
3. Solve this fully within `JsonSerializer`, by using `beginReadDictionaryEntry`/`endReadDictionaryEntry` as a signal that the value has been consumed (requires something like a `bool* m_dictionaryEntryConsumed` field that points to a `bool` within `readDictionary`'s stack frame and treated as `true` if it has been set to a different address after `entry_callback` has returned, to support nested dictionaries)

`3.` would obviously be the least intrusive one, but having `2.` available might be handy for other serializers, too.

**Comment by HenkKalkwater** [comment]
As you mentioned, I indeed didn't take API compatibility into account, perhaps since I assumed it wasn't an issue due to vibe-d still being under version 1.0.

I'll change my current implementation as mentioned in option 2. I'm not a huge fan of option 3, I believe it would complicate the execution flow of the code too much, making debugging harder. And I'm certain my implementation won't be perfect directly after writing it.

File: data/vibe/data/serialization.d [hunk 1]
```diff
@@ -726,7 +731,11 @@ private template deserializeValueImpl(Serializer, alias Policy) {
 				bool[fieldsCount] set;
 				ser.readDictionary!Traits((name) {
 					switch (name) {
-						default: break;
+						default:
+							static if(is(typeof(ser.skipvalue))) {
```
Comment (s-ludwig) [inline data/vibe/data/serialization.d:735]:
```suggestion
							static if (is(typeof(ser.skipValue()))) {
```

File: data/vibe/data/serialization.d [hunk 2]
```diff
@@ -888,7 +897,11 @@ private template deserializeValueImpl(Serializer, alias Policy) {
 				ser.readDictionary!Traits((name) {
 					static if (hasSerializableFields!(T, Policy)) {
 						switch (name) {
-							default: break;
+							default:
+								static if(is(typeof(ser.skipValue))) {
```
Comment (s-ludwig) [inline data/vibe/data/serialization.d:901]:
```suggestion
								static if (is(typeof(ser.skipValue()))) {
```

Comment (HenkKalkwater) [reply data/vibe/data/serialization.d:901]: That would make more sense, yes.

**Review by s-ludwig — APPROVED** [review]

**Comment by HenkKalkwater** [comment]
This should also have closed #1961 I believe.

**Pull request merged**